core/new-gui/src/app/workspace/service/workflow-graph/model/shared-model-change-handler.ts

```diff
@@ -0,0 +1,343 @@
+import { WorkflowGraph } from "./workflow-graph";
+import { JointGraphWrapper } from "./joint-graph-wrapper";
+import * as Y from "yjs";
+import {
+  Breakpoint,
+  CommentBox,
+  OperatorLink,
+  OperatorPredicate,
+  Point,
+} from "../../../types/workflow-common.interface";
+import { JointUIService } from "../../joint-ui/joint-ui.service";
+import * as joint from "jointjs";
+import { environment } from "../../../../../environments/environment";
+import { YType } from "../../../types/shared-editing.interface";
+
+/**
+ * SyncJointModelService listens to changes to the TexeraGraph (SharedModel) and updates Joint graph correspondingly,
+ * regardless of whether the changes are local or from other co-editors.
+ *
+ * In all the handlers, this service will also trigger the necessary subjects, which were written in
+ * <code>{@link WorkflowGraph}</code> in previous implementations and now all migrated here.
+ *
+ */
+
+export class SharedModelChangeHandler {
+  constructor(
+    private texeraGraph: WorkflowGraph,
+    private jointGraph: joint.dia.Graph,
+    private jointGraphWrapper: JointGraphWrapper,
+    private jointUIService: JointUIService
+  ) {
+    this.handleOperatorAddAndDelete();
+    this.handleLinkAddAndDelete();
+    this.handleElementPositionChange();
+    this.handleCommentBoxAddAndDelete();
+    this.handleBreakpointAddAndDelete();
+    this.handleOperatorDeep();
+    this.handleCommentBoxDeep();
+    this.texeraGraph.newYDocLoadedSubject.subscribe(_ => {
+      this.handleOperatorAddAndDelete();
```

Comment:
when a new y doc is loaded, do we need to call `unobserve` to `Unregister` the observer functions on the previous ydoc object? this might be a source of memory leaks

Reply:
We don't need to according to the [documentation](https://docs.yjs.dev/api/y.doc) for `yDoc.destroy()`.